### ISSUE TITLE
Package rename introduces .bin conflict

### ISSUE THREAD

**Issue by jkrems** [jashkenas/coffeescript#4810]
Due to the recent renaming, there's now two popular packages trying to install `node_modules/.bin/coffee`. This means that if any two packages disagree on which of the two to install (and flattening the package tree), there's suddenly a race condition which determines who gets the symlink. Depending on the package manager, the install will just fail.

**Comment by GeoffreyBooth** [comment]
Steps to reproduce?

**Comment by jkrems** [comment]
A short one is:

```
yarn add coffee-script coffeescript
```

It's now random (?) what `node_modules/.bin/coffee` will point to.

**Comment by GeoffreyBooth** [comment]
It’s actually not limited to `coffee-script` versus `coffeescript`. You can get the same conflict with dependencies using different versions of the `coffeescript` package:

```bash
▶ mkdir test && cd test
▶ npm install gulp-coffee  # gulp-coffee depends on "coffeescript": "^1.10.0"
+ gulp-coffee@2.3.5
added 60 packages in 3.505s
▶ ls -hal node_modules/.bin
total 16
drwxr-xr-x   4 geoffrey  wheel   136B Dec  1 14:30 .
drwxr-xr-x  60 geoffrey  wheel   2.0K Dec  1 14:30 ..
lrwxr-xr-x   1 geoffrey  wheel    24B Dec  1 14:30 cake -> ../coffeescript/bin/cake
lrwxr-xr-x   1 geoffrey  wheel    26B Dec  1 14:30 coffee -> ../coffeescript/bin/coffee
▶ npx coffee -v
CoffeeScript version 1.12.7
▶ npm i coffeescript
+ coffeescript@2.0.3
added 1 package and updated 1 package in 1.23s
▶ npx coffee -v
CoffeeScript version 2.0.3
```

**Comment by jkrems** [comment]
The same happens when babel 6 split into `babel-core` and `babel-cli`. Any `babel@5` in the dependency tree was enough to ruin people's days.

**Comment by GeoffreyBooth** [comment]
The symlinks in `node_modules/.bin` get put there because of these lines in CoffeeScript’s `package.json`:

```
  "bin": {
    "coffee": "./bin/coffee",
    "cake": "./bin/cake"
  },
```

So really this is a package manager issue, as there’s no code of ours creating those symlinks. NPM or Yarn need to intelligently deconflict `bin` keys between various packages.

Even if we took out the `bin` key going forward, that wouldn’t solve the issue for all the versions that have already been released, that we can’t modify. Similarly, we could add a `postinstall` script (shudder) that would try to ensure that the symlinks point to the newest `coffeescript` package, but that would similarly get overridden if some older `coffeescript` package gets installed later (such as by a dependency like `gulp-coffee`). Ditto if we edited the code of what the `coffee` symlink points to, as the later install would be overwriting the symlink to point to the older `coffee` binary that would lack such code.

So basically, since we can’t edit already-released versions, there’s really nothing we can do about this. It’s up to NPM and Yarn to solve. If I were them, in the case of two packages trying to install the same binary or symlink, I would decide between them based on:
- Prefer “top-level” packages, i.e. `coffeescript` in the user’s `package.json` over `coffeescript` installed as a dependency of some other package;
- Prefer non-deprecated packages over deprecated ones, which would prefer `coffeescript` over `coffee-script` if both are top-level ([once we deprecate `coffee-script`](https://github.com/jashkenas/coffeescript/issues/4714#issuecomment-348630870)).

If NPM and/or Yarn implement these changes, the issue goes away. I would suggest you open issues with those projects.

**Comment by jkrems** [comment]
This is a known issue whenever 2 packages expose the same binary and unlikely to be addressed by yarn or npm. :( Just wanted to open it here early so people will find it. For our own uses we'll just pin `coffee-script` at a version before the huge warning was introduced. Given our uses of coffee, I don't think the pain of pulling in the rename will be worth it for us. :)

**Comment by GeoffreyBooth** [comment]
The warning is gone and I’ve unpublished those versions. It shouldn’t be reachable anymore.

**Comment by jkrems** [comment]
Given that the bin conflict exists and keeps getting worse by creating a wild mix on npm right now, we'll have to double down on our efforts to remove CoffeeScript from all our projects. It's messing too much with the stability of our builds.

**Comment by GeoffreyBooth** [comment]
NPM doesn’t allow deleting or unpublishing old packages or versions after 24 hours. I appreciate your frustration, but there’s nothing I can do to help you.

Aside from complaining to NPM, you might want to try setting up your own NPM registry via something like https://github.com/rlidwka/sinopia and configuring your projects to use that instead of the public registry. Then you can create overrides, for example to make `coffee-script` an alias for `coffeescript`. See https://github.com/rlidwka/sinopia#override-public-packages

**Comment by jkrems** [comment]
It would have been possible to quickly deprecate `coffeescript` and go back to the old name. We're already running our own registry proxy but we'd have actively patch incoming package.json files if we'd want to get rid of the bin collision. The problem is the file layout after already being installed by npm or yarn.

Anyhow, we'll work around with a proxy package on disk that also allows us to print deprecation warnings whenever someone requires coffee-script. Hopefully that will get rid of this problem soon.
